### PR TITLE
Optimize reading MobiHeader

### PR DESCRIPTION
**Pull request by philippeitis**
We don't need to read all the bytes in order to read the mobi metadata, but this is what currently happens.

If we only read the bytes we need, we get a 10x speedup (times averaged over a set of >100 files using MobiMetadata::from_path):

```
old MOBI 0.0009397881355932203
new MOBI: 0.00010762542372881359
```

The major change is switching the Reader to a MobiReader trait, which we can then use to allow the original Reader implementation, and a new one (ReaderPrime, but this can be renamed), which doesn't need to have all the bytes stored in memory and is accordingly a lot faster.

The code is a bit unideal (storing results of buffer inside a buffering vec), but the speed up is well worth it, and I haven't noticed any weird bugs when using it in my application.

### CODE REVIEW

**Comment by wojciechkepka** [comment]
In general the speedups and changes look great, I'm starting to wonder whether we actually need the original reader if this one has the same functionality.

**Comment by philippeitis** [comment]
Still need the Reader to avoid copying the contents of Vec<u8>. A better implementation would strictly read the files in a sequential fashion, which allows us to avoid maintaining a copy of the data twice (eg. in Reader and in Mobi), and would make things a lot nicer.

**Comment by philippeitis** [comment]
In order to ensure sequential reads, I decided to create a Records class, which stores all record offsets and ids, and parse that from the beginning (although I'm not really happy about doing unnecessary work - this could be fixed by making MobiMetadata as an external class, MobiMetadataWithRecords as an internal class, and adding a conversion?). However, this allows us to avoid going backwards in the reader, so we should be able to just use a Reader and not maintain data internally.

**Comment by philippeitis** [comment]
I've managed to make the reads fully sequential, and I've also discovered that MobiHeader reads its name from the first record.

It might be a good idea to have this name be read into a separate struct (possibly reading any relevant data from the first record as well), so that we don't have one header reading bytes from the next header.

**Comment by wojciechkepka** [comment]
So while I don't have as much time to dive deeper into this context right now from what I've gathered it might be a good idea to get rid of name from MobiHeader. As you've pointed out it doesn't technically belong there. Do you have any information on what else is in record 0? If it is the case that there are more useful fields than I'm all for adding a new wrapper struct for the record.

File: src/headers/mobih.rs [hunk 1]
```diff
@@ -106,46 +101,32 @@ impl MobiHeader {
...
+            name: String::new(),
         })
     }
 
-    /// Returns the book name
-    pub(crate) fn name(reader: &mut Reader) -> io::Result<String> {
-        let name_offset = reader.read_u32_header(MobiHeaderData::NameOffset)?;
-        let name_length = reader.read_u32_header(MobiHeaderData::NameLength)?;
+    pub(crate) fn finish_parse(&mut self, reader: &mut impl MobiReader) -> io::Result<()> {
         // TODO: figure out why is this exactly `+ 80` and it works?
```
Comment (philippeitis) [inline src/headers/mobih.rs:116]: Header is 78 bytes, records is 8 * num_records bytes + 2 padding bytes, and then the first record starts at the end of the list of records, so we do 80 + num_records * 8 + offset to get the offset in the first record.

Comment (wojciechkepka) [reply src/headers/mobih.rs:116]: That makes sense. I guess I figured it out at some point but forgot to make a note.

**Comment by wojciechkepka** [comment]
I've found this:
https://wiki.mobileread.com/wiki/MOBI
```
Remainder of Record 0

At the end of Record 0 of the PDB file format, we usually get the full file name, the offset of which is given in the MOBI header.

There might be data of unknown use between the end of the EXTH records and the name.

The name is followed by two null bytes, and then padded with null bytes to a four-byte boundary. For example, if the name is 16 bytes long, with two null bytes, that makes 18 bytes, and it then gets another two null bytes added to make it up to 20 bytes in total. However, the length stored in the header is only 16. If the name was 19 bytes, it would be followed by two null bytes to make it up to 21 bytes, and then padded with three more null bytes to make it up to 24 bytes.

The name and padding is followed by more data of unknown use, usually null bytes, to the end of section 0. 
```

File: src/headers/header.rs [hunk 2]
```diff
@@ -42,14 +44,14 @@ pub struct Header {
...
 }
 impl Header {
     /// Parse a header from the content
-    pub(crate) fn parse(reader: &mut Reader) -> io::Result<Header> {
+    pub(crate) fn parse(reader: &mut impl MobiReader) -> io::Result<Header> {
         use HeaderData::*;
         Ok(Header {
-            name: reader.read_string_header(Name, 32),
+            name: reader.read_string_header(Name.position(), 32)?,
             attributes: reader.read_i16_header(Attributes)?,
```
Comment (philippeitis) [inline src/headers/header.rs:55]: Why are attributes and version i16s? They seem like they should be u16?

Comment (wojciechkepka) [reply src/headers/header.rs]: They should be u16.

It's a bitfield with this values:
 0x0002 Read-Only 
 0x0004 Dirty AppInfoArea 
 0x0008 Backup this database (i.e. no conduit exists) 
 0x0010 (16 decimal) Okay to install newer over existing copy, if present on PalmPilot 
 0x0020 (32 decimal) Force the PalmPilot to reset after this database is installed 
 0x0040 (64 decimal) Don't allow copy of file to be beamed to other Pilot.

Seems like read-only is something worth extracting.

**Comment by philippeitis** [comment]
I think it's a good idea to get rid of HeaderField entirely, since it's really confusing to think about (eg. how does HeaderField relate to the offsets in the start Header / records? Why do we use num_records in one place and not another? Where is it coming from?) and instead read each header in sequence, since they appear after each other anyways.

This simplifies the code a lot and makes reasoning about the individual headers many, many times easier, since we know that they appear in order how they'd be parsed individually.

**Comment by wojciechkepka** [comment]
Thats a great idea. Sequential reading is definitely the way to go. So your making great points about num_records, its not really documented in the current state and your implementation is much clearer and more readable.

I'm a beginner at programming and this was one of my first attempts at writing something bigger so there might be some pieces that will stand out as a typical noobs mistake but your pull requests have been a great source of learning so far so thanks for that :)

**Comment by wojciechkepka** [comment]
Is there anything you'd like to add here or should I merge it?

**Comment by philippeitis** [comment]
Since Reader was only being used in tests, I removed it and the trait.

I think that we could make Mobi look a bit like this:

```
enum RecordState<R> {
    Unread(Reader<R>),
    Read(Vec<Record>)
}

/// Structure that holds parsed ebook information and contents
pub struct Mobi<R> {
    record_state: RecordState<R>, 
    pub metadata: MobiMetadata,
}

impl Mobi<R> {
    /// Returns all readable records content decompressed as a Vec
    pub fn content(&self) -> io::Result<Vec<u8>> {
        match self.record_state {
            ...
        }
}
```

This way, if we need the records, we can process them once and hold on to a copy internally, but if we don't, we just hold on to the reader until they're either needed or the mobi is dropped - this would have the benefit of making Mobi and MobiMetadata equally fast for processing metadata.

**Comment by philippeitis** [comment]
I decided to move the name field from MobiHeader to MobiMetadata, since name doesn't actually exist in the bytes that MobiMetadata has access to (eg. from end of palmdoc header to start of ext header) - but Mobi does, and we can even make the exact behaviour really obvious (since it's the data in record 0, at the specified name offset), which I think is important for clarity.

Every header now correctly takes all the bytes it should, which makes testing the individual components really easy, since we can pass the exact bytes for the particular header we want to check.

As an additional note, reading extheaders should be more efficient since we're reading all the bytes we need in a vectorized fashion.

Reading metadata is now ~25x faster :) 

```
old MOBI 0.0009397881355932203
new MOBI: 3.942627118644066e-05
```

**Comment by wojciechkepka** [comment]
Nice! All those changes together really bring in some speed ups :) Not only is the speed nice but reasoning about the code is much easier and the code structure will allow for more tests. Lots of magic is now gone from the code and it feels a lot more robust (more like a crate should actually be).

**Pull request merged**